pkg/agent/cniserver/pod_monitor.go

```diff
@@ -0,0 +1,208 @@
+				// Update Pod OpenFlow entries only after the OpenFlow port state is live.
+				if status.Desc.State == openflow15.PS_LIVE {
+					m.updateUnreadyPod(status)
+				}
+			case <-ticker:
+				m.checkUnreadyPods()
+			}
+		}
+	}()
+}
+
+func (m *podIfaceMonitor) updatePodFlows(ifName string, ofPort int32, podInfo *unreadyPodInfo) error {
+	ifConfig, found := m.ifaceStore.GetInterfaceByName(ifName)
+	if !found {
+		klog.InfoS("Interface config is not found in local cache, remove from local cache", "port", ifName)
+		delete(m.unreadyInterfaces, ifName)
+		return nil
+	}
+
+	// Update interface config with the ofPort.
+	ifConfig.OVSPortConfig.OFPort = ofPort
+	m.ifaceStore.UpdateInterface(ifConfig)
+
+	// Install OpenFlow entries for the Pod.
+	klog.V(2).InfoS("Setting up Openflow entries for OVS port", "port", ifName)
+	if err := m.ofClient.InstallPodFlows(ifName, ifConfig.IPs, ifConfig.MAC, uint32(ofPort), ifConfig.VLANID, nil); err != nil {
+		return fmt.Errorf("failed to add Openflow entries for OVS port %s: %v", ifName, err)
+	}
+
+	// Notify the Pod update event to required components.
+	event := agenttypes.PodUpdate{
+		PodName:      ifConfig.PodName,
+		PodNamespace: ifConfig.PodNamespace,
+		IsAdd:        true,
+		ContainerID:  ifConfig.ContainerID,
+	}
+	m.podUpdateNotifier.Notify(event)
+
+	podInfo.flowInstalled = true
+	// Update Pod event immediately
+	m.processPodEvents(ifName, podInfo)
+	return nil
+}
+
+func (m *podIfaceMonitor) updateUnreadyPod(status *openflow15.PortStatus) {
+	ovsPort := string(bytes.Trim(status.Desc.Name, "\x00"))
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	podInfo, found := m.unreadyInterfaces[ovsPort]
+	if !found {
+		klog.InfoS("OVS port is not found", "ovsPort", ovsPort)
+		return
+	}
+
+	ofPort := status.Desc.PortNo
+	if err := m.updatePodFlows(ovsPort, int32(ofPort), podInfo); err != nil {
+		klog.ErrorS(err, "Failed to update Pod's OpenFlow entries, requeue the PortStatus message after 10s", "PodName", podInfo.podName, "PodNamespace", podInfo.podNamespace, "OVSPort", ovsPort)
+		m.reQueuePortStatus(status)
+		return
+	}
+}
+
+func (m *podIfaceMonitor) checkUnreadyPods() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for key, podInfo := range m.unreadyInterfaces {
+		m.processPodEvents(key, podInfo)
+	}
+}
+
+func (m *podIfaceMonitor) processPodEvents(ovsPort string, podInfo *unreadyPodInfo) {
+	// Ignore the un-ready Pod if time is not up.
+	if !podInfo.flowInstalled && time.Now().Sub(podInfo.createTime).Seconds() < podNotReadyTimeInSeconds {
+		return
+	}
+
+	pod, err := m.podLister.Pods(podInfo.podNamespace).Get(podInfo.podName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get Pod, retrying", "Namespace", podInfo.podNamespace, "Name", podInfo.podName)
+		return
+	}
+
+	if podInfo.flowInstalled {
+		// Add normal event to record Pod network is ready.
+		m.record.Eventf(pod, corev1.EventTypeNormal, "NetworkIsReady", "Installed Pod '%s/%s' network forwarding rules", podInfo.podNamespace, podInfo.podName)
```

Comment:
@tnqn since recording an event is non-blocking, it's possible that this event will not be published (e.g. if antrea-agent stops). I assume this is not an issue, the "warning" events will become stale after some time and recording a "normal" event here is just a nice-to-have?

Reply:
Right, the generation of events are for visibility only. I guess the possibility of losing events on restart applies to K8s components too.